src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -15,6 +17,11 @@ namespace System.Text.Unicode
 {
     internal static unsafe partial class Utf8Utility
     {
+        private static readonly Vector128<byte> s_mostSignficantBitMask = Vector128.Create((byte)0x80);
```

Comment:
I have seen in the past that referring `static` variable is slower on ARM64 because it has to call helper. Calling helper is done by loading its address which is 3 instructions + a helper call. Since you will be accessing it inside a loop, I would recommend to make these local variables. Here is a quick test:

```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static int AdvSimdMoveMask1(Vector128<byte> value)
{
    Debug.Assert(AdvSimd.Arm64.IsSupported);

    // extractedBits[i] = (value[i] & 0x80) == 0x80 & (1 << i);
    Vector128<byte> mostSignficantBitMask = s_mostSignficantBitMask;
    Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, mostSignficantBitMask), mostSignficantBitMask);
    Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitMask128);

    // self-pairwise add until all flags have moved to the first two bytes of the vector
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    return extractedBits.AsInt32().ToScalar();
}
```

<details>

<summary>Generated code for AdvSimdMoveMask1</summary>

```asm
; Assembly listing for method BitArrayTest.TestClass:Main(System.String[]):int
; Emitting BLENDED_CODE for generic ARM64 CPU - Windows
; optimized code
; fp based frame
; partially interruptible
; Final local variable assignments
;
;* V00 arg0         [V00    ] (  0,  0   )     ref  ->  zero-ref    class-hnd
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [sp+0x00]   "OutgoingArgSpace"
;  V02 tmp1         [V02,T05] (  3,  3   )  simd16  ->  d16         HFA(simd16)  "Inline stloc first use temp"
;  V03 tmp2         [V03,T04] (  2,  4   )  simd16  ->   d8         HFA(simd16)  "Inlining Arg"
;  V04 tmp3         [V04,T01] (  3,  6   )  simd16  ->  d16         HFA(simd16)  "dup spill"
;  V05 tmp4         [V05,T02] (  3,  6   )  simd16  ->  d16         HFA(simd16)  "dup spill"
;  V06 tmp5         [V06,T03] (  3,  6   )  simd16  ->  d16         HFA(simd16)  "dup spill"
;* V07 cse0         [V07,T00] (  0,  0   )    long  ->  zero-ref    "CSE - aggressive"
;
; Lcl frame size = 0

G_M58909_IG01:
        A9BE7BFD          stp     fp, lr, [sp,#-32]!
        6D0127E8          stp     d8, d9, [sp,#16]
        910003FD          mov     fp, sp
                                                ;; bbWeight=1    PerfScore 2.50
G_M58909_IG02:
        4F00E5E8          movi    v8.16b, #0x0f
        D291BF00          movz    x0, #0x8df8
        F2BE2860          movk    x0, #0xf143 LSL #16
        F2CFFF60          movk    x0, #0x7ffb LSL #32
        52800021          mov     w1, #1
        6E084509          mov     v9.d[0], v8.d[1]
        97FF44AF          bl      CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
        D2858700          movz    x0, #0x2c38
        F2B8D1C0          movk    x0, #0xc68e LSL #16
        F2C03BC0          movk    x0, #478 LSL #32
        F9400000          ldr     x0, [x0]
        3CC08010          ldr     q16, [x0,#8]
        6E180528          mov     v8.d[1], v9.d[0]
        4E301D11          and     v17.16b, v8.16b, v16.16b
        6E308E30          cmeq    v16.16b, v17.16b, v16.16b
        D2858800          movz    x0, #0x2c40
        F2B8D1C0          movk    x0, #0xc68e LSL #16
        F2C03BC0          movk    x0, #478 LSL #32
        F9400000          ldr     x0, [x0]
        3CC08011          ldr     q17, [x0,#8]
        4E311E10          and     v16.16b, v16.16b, v17.16b
        4E30BE10          addp    v16.16b, v16.16b, v16.16b
        4E30BE10          addp    v16.16b, v16.16b, v16.16b
        4E30BE10          addp    v16.16b, v16.16b, v16.16b
        4E042E00          smov    x0, v16.s[0]
                                                ;; bbWeight=1    PerfScore 29.50
G_M58909_IG03:
        6D4127E8          ldp     d8, d9, [sp,#16]
        A8C27BFD          ldp     fp, lr, [sp],#32
        D65F03C0          ret     lr
                                                ;; bbWeight=1    PerfScore 3.00

; Total bytes of code 124, prolog size 12, PerfScore 47.40, (MethodHash=ba8f19e2) for method BitArrayTest.TestClass:Main(System.String[]):int
; ============================================================
```
</details>

vs.

```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static int AdvSimdMoveMask2(Vector128<byte> value, Vector128<byte> mostSignficantBitMask, Vector128<byte> s_bitMask128)
{
    Debug.Assert(AdvSimd.Arm64.IsSupported);

    // extractedBits[i] = (value[i] & 0x80) == 0x80 & (1 << i);
    Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, mostSignficantBitMask), mostSignficantBitMask);
    Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitMask128);

    // self-pairwise add until all flags have moved to the first two bytes of the vector
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    return extractedBits.AsInt32().ToScalar();
}
```

<details>
<summary>Generated code for AdvSimdMoveMask2</summary>
```asm
; Assembly listing for method BitArrayTest.TestClass:Main(System.String[]):int
; Emitting BLENDED_CODE for generic ARM64 CPU - Windows
; optimized code
; fp based frame
; partially interruptible
; Final local variable assignments
;
;* V00 arg0         [V00    ] (  0,  0   )     ref  ->  zero-ref    class-hnd
;  V01 loc0         [V01,T04] (  3,  3   )  simd16  ->  d16         HFA(simd16)
;* V02 loc1         [V02    ] (  0,  0   )  simd16  ->  zero-ref    HFA(simd16)
;# V03 OutArgs      [V03    ] (  1,  1   )  lclBlk ( 0) [sp+0x00]   "OutgoingArgSpace"
;  V04 tmp1         [V04,T05] (  2,  2   )  simd16  ->  d17         HFA(simd16)
;  V05 tmp2         [V05,T03] (  2,  4   )  simd16  ->  d18         HFA(simd16)  "Inlining Arg"
;* V06 tmp3         [V06    ] (  0,  0   )  simd16  ->  zero-ref    HFA(simd16)  "Inlining Arg"
;* V07 tmp4         [V07    ] (  0,  0   )  simd16  ->  zero-ref    HFA(simd16)  "Inlining Arg"
;  V08 tmp5         [V08,T00] (  3,  6   )  simd16  ->  d16         HFA(simd16)  "dup spill"
;  V09 tmp6         [V09,T01] (  3,  6   )  simd16  ->  d16         HFA(simd16)  "dup spill"
;  V10 tmp7         [V10,T02] (  3,  6   )  simd16  ->  d16         HFA(simd16)  "dup spill"
;
; Lcl frame size = 0

G_M58909_IG01:
        A9BF7BFD          stp     fp, lr, [sp,#-16]!
        910003FD          mov     fp, sp
                                                ;; bbWeight=1    PerfScore 1.50
G_M58909_IG02:
        4F04E410          movi    v16.16b, #0x80
        D2804020          movz    x0, #513
        F2A10080          movk    x0, #0x804 LSL #16
        F2C40200          movk    x0, #0x2010 LSL #32
        F2F00800          movk    x0, #0x8040 LSL #48
        4E080C11          dup     v17.2d, x0
        4F00E5F2          movi    v18.16b, #0x0f
        4E301E52          and     v18.16b, v18.16b, v16.16b
        6E308E50          cmeq    v16.16b, v18.16b, v16.16b
        4E311E10          and     v16.16b, v16.16b, v17.16b
        4E30BE10          addp    v16.16b, v16.16b, v16.16b
        4E30BE10          addp    v16.16b, v16.16b, v16.16b
        4E30BE10          addp    v16.16b, v16.16b, v16.16b
        4E042E00          smov    x0, v16.s[0]
                                                ;; bbWeight=1    PerfScore 14.00
G_M58909_IG03:
        A8C17BFD          ldp     fp, lr, [sp],#16
        D65F03C0          ret     lr
                                                ;; bbWeight=1    PerfScore 2.00

; Total bytes of code 72, prolog size 8, PerfScore 24.70, (MethodHash=ba8f19e2) for method BitArrayTest.TestClass:Main(System.String[]):int
; ============================================================
```
</details>

Reply:
Does the reduced instruction count justify the added complexity of hoisting what amounts to implementation detail to the callsite?

Reply:
For these hot paths, probably. But it also is something we should likely explicitly track as the JIT should realistically be hoisting this itself.

Reply:
JIT won't hoist access to static variable because it is not an invariant. Different thread can change its value in middle of loop. See https://github.com/dotnet/runtime/issues/35279#issuecomment-620125815

Reply:
Perhaps I'm missing something

> This all changes for readonly statics since they can't be safely modified after their corresponding class constructors have completed.

If the JIT doesn't optimize readonly statics, then surely it should?

Reply:
Yes, `readonly` static optimize the helper call. I didn't realize that those are `readonly static`. I tried it on my local machine and looks like `s_bitMask128` variable access is getting hoisted. However `s_mostSignficantBitMask` is not getting hoisted. Not saving it to a local variable `Vector128<byte> mostSignficantBitMask = s_mostSignficantBitMask;` makes its access hoisted.



```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static ulong AllStatic(Vector128<byte> value)
{
    Debug.Assert(AdvSimd.Arm64.IsSupported);

    Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, s_mostSignficantBitMask), s_mostSignficantBitMask);
    Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitMask128);
    // self-pairwise add until all flags have moved to the first two bytes of the vector
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    ulong result = extractedBits.AsUInt64().ToScalar();
    return result;
}
```

```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static ulong OneStatic(Vector128<byte> value)
{
    Debug.Assert(AdvSimd.Arm64.IsSupported);

    Vector128<byte> mostSignficantBitMask = s_mostSignficantBitMask;
    Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, s_mostSignficantBitMask), s_mostSignficantBitMask);
    Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitMask128);
    // self-pairwise add until all flags have moved to the first two bytes of the vector
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    ulong result = extractedBits.AsUInt64().ToScalar();
    return result;
}
```

```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static ulong NoStatic(Vector128<byte> value, Vector128<byte> mostSignficantBitMask, Vector128<byte> bitMask128)
{
    Debug.Assert(AdvSimd.Arm64.IsSupported);

    Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, mostSignficantBitMask), mostSignficantBitMask);
    Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, bitMask128);
    // self-pairwise add until all flags have moved to the first two bytes of the vector
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    ulong result = extractedBits.AsUInt64().ToScalar();
    return result;
}
```
NoStatic vs. OneStatic

![image](https://user-images.githubusercontent.com/12488060/87175848-16b00380-c28e-11ea-850a-0b64b073ee5d.png)

NoStatic vs. AllStatic

![image](https://user-images.githubusercontent.com/12488060/87175899-26c7e300-c28e-11ea-8405-2013fa4d4d7a.png)

Summarize: 
I need to look deeper on why OneStatic version doesn't CSE the load, but for now, probably don't store `s_mostSignficantBitMask` in local variable, but just use it directly in `AdvSimd.And()`.

@briansull - FYI

Reply:
Interesting, thanks for the analysis!

Reply:
As the CSE decision might differ based on other factors in the method, you should compare the JIT code emitted with and without static variables and choose the optimal one. You can dump the JIT code by setting environment variable `set COMPlus_JITDisasm=GetPointerToFirstInvalidByte`.